File: skills/sortie/lib/parse_jsonl_metrics.py
"""
Parse Claude Code JSONL session logs to extract agent activity metrics.

Session files live at:
    ~/.claude/projects/<encoded-path>/<session-uuid>.jsonl
    ~/.claude/projects/<encoded-path>/<session-uuid>/subagents/*.jsonl

Path encoding: replace all '/' and '.' with '-'.
Each line is a JSON record. We care about:
    - type:"assistant" — contains tool_use blocks (tool calls) and usage data
    - type:"user"      — contains tool_result blocks (errors when is_error=True)
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional

CLAUDE_PROJECTS_DIR = Path.home() / ".claude" / "projects"


def _safe_int(v: object) -> int:
    """Safely coerce a potentially-malformed token count value to an integer.

    Mirrors JS behaviour: parse as float first, then truncate — so "1.9" -> 1.
    """
    try:
        if v is None:
            return 0
        n = float(v)  # type: ignore[arg-type]
        return int(n) if math.isfinite(n) else 0
    except (TypeError, ValueError):
        return 0


def _safe_mtime(p: Path) -> float:
    """Return mtime for sorting; returns 0.0 on OSError (race with deletion)."""
    try:
        return p.stat().st_mtime
    except OSError:
        return 0.0


def encode_project_path(abs_path: str) -> str:
    """Encode an absolute path the same way Claude Code does for its project dirs.

    Both '/' and '.' are replaced with '-'.
    """
    return abs_path.replace("/", "-").replace(".", "-")


def _find_all_jsonl_files(directory: Path) -> List[Path]:
    """Recursively find all .jsonl files under directory (for subagent logs)."""
    files: List[Path] = []
    try:
        for entry in directory.iterdir():
            if entry.is_file() and entry.suffix == ".jsonl":
                files.append(entry)
            elif entry.is_dir():
                files.extend(_find_all_jsonl_files(entry))
    except OSError:
        pass
    return files


def find_latest_session_file(worktree_path: str) -> Optional[Path]:
    """Find the most recently modified top-level JSONL session file for a given worktree path."""
    encoded = encode_project_path(worktree_path)
    project_dir = CLAUDE_PROJECTS_DIR / encoded

    if not project_dir.is_dir():
        return None

    jsonl_files = list(project_dir.glob("*.jsonl"))
    if not jsonl_files:
        return None

    # Pick the most recently modified (safe against deletion races)
    jsonl_files.sort(key=_safe_mtime, reverse=True)
    return jsonl_files[0]


@dataclass
class JsonlMetrics:
    session_id: Optional[str] = None
    session_file: str = ""
    tool_call_counts: Dict[str, int] = field(default_factory=dict)
    total_tool_calls: int = 0
    error_count: int = 0
    error_rate: float = 0.0
    agent_spawns: int = 0
    last_activity_at: Optional[str] = None
    recent_timeline: List[Dict[str, str]] = field(default_factory=list)
    # Token usage (aggregated across all sessions for the worktree)
    input_tokens: int = 0
    output_tokens: int = 0
    cache_write_tokens: int = 0
    cache_read_tokens: int = 0
    total_tokens: int = 0


def _parse_single_file(path: Path, tool_call_counts: Dict[str, int],
                       timeline: List[Dict[str, str]],
                       accumulator: Dict) -> None:
    """Parse one JSONL file, mutating the shared accumulators."""
    try:
        with path.open(encoding="utf-8", errors="replace") as fh:
            for line in fh:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if accumulator["session_id"] is None and obj.get("sessionId"):
                    accumulator["session_id"] = obj["sessionId"]

                entry_type = obj.get("type")

                if entry_type == "assistant":
                    # Token usage is on the message object; coerce to int to guard
                    # against malformed values (strings, None, floats) in JSONL.
                    usage = obj.get("message", {}).get("usage") or {}
                    accumulator["input_tokens"]       += _safe_int(usage.get("input_tokens"))
                    accumulator["output_tokens"]      += _safe_int(usage.get("output_tokens"))
                    accumulator["cache_write_tokens"] += _safe_int(usage.get("cache_creation_input_tokens"))
                    accumulator["cache_read_tokens"]  += _safe_int(usage.get("cache_read_input_tokens"))

                    content = obj.get("message", {}).get("content", [])
                    if isinstance(content, list):
                        for block in content:
                            if (
                                isinstance(block, dict)
                                and block.get("type") == "tool_use"
                            ):
                                name = block.get("name") or "unknown"
                                tool_call_counts[name] = (
                                    tool_call_counts.get(name, 0) + 1
                                )
                                if name == "Agent":
                                    accumulator["agent_spawns"] += 1
                                timeline.append(
                                    {
                                        "tool": name,
                                        "timestamp": obj.get("timestamp"),
                                    }
                                )

                elif entry_type == "user":
                    content = obj.get("message", {}).get("content", [])
                    if isinstance(content, list):
                        for block in content:
                            if (
                                isinstance(block, dict)
                                and block.get("type") == "tool_result"
                                and block.get("is_error") is True
                            ):
                                accumulator["error_count"] += 1
    except OSError:
        pass


def parse_jsonl_metrics(worktree_path: str) -> Optional[JsonlMetrics]:
    """Load JSONL metrics for a worktree path.

    Scans ALL session files (including subagent logs) for token usage.
    Returns None if no session files exist.
    """
    encoded = encode_project_path(worktree_path)
    project_dir = CLAUDE_PROJECTS_DIR / encoded

    if not project_dir.is_dir():
        return None

    all_files = _find_all_jsonl_files(project_dir)
    if not all_files:
        return None

    # Latest top-level session for metadata (session_id, session_file)
    top_level = [f for f in all_files if f.parent == project_dir]
    latest_session_file: Optional[Path] = None
    if top_level:
        top_level.sort(key=_safe_mtime, reverse=True)
        latest_session_file = top_level[0]

    tool_call_counts: Dict[str, int] = {}
    timeline: List[Dict[str, str]] = []
    acc = {
        "session_id": None,
        "input_tokens": 0,
        "output_tokens": 0,
        "cache_write_tokens": 0,
        "cache_read_tokens": 0,
        "agent_spawns": 0,
        "error_count": 0,
        "last_activity_at": None,
    }

    # Parse latest session file first so session_id comes from the correct session
    if latest_session_file:
        _parse_single_file(latest_session_file, tool_call_counts, timeline, acc)
    for path in all_files:
        if path == latest_session_file:
            continue
        _parse_single_file(path, tool_call_counts, timeline, acc)

    # Sort timeline by timestamp so cross-file ordering is correct, then derive
    # last_activity_at from the true latest event rather than traversal order.
    # Parse to epoch so ISO 8601 timestamps with offsets sort correctly.
    def _ts_epoch(ts: Optional[str]) -> float:
        if not ts:
            return 0.0
        try:
            return datetime.fromisoformat(ts.replace("Z", "+00:00")).timestamp()
        except (ValueError, AttributeError):
            return 0.0

    timeline.sort(key=lambda e: _ts_epoch(e.get("timestamp")))
    if timeline:
        acc["last_activity_at"] = timeline[-1].get("timestamp")

    total_tool_calls = sum(tool_call_counts.values())
    error_rate = (
        round(acc["error_count"] / total_tool_calls, 2) if total_tool_calls > 0 else 0.0
    )
    total_tokens = (
        acc["input_tokens"] + acc["output_tokens"] +
        acc["cache_write_tokens"] + acc["cache_read_tokens"]
    )

    return JsonlMetrics(
        session_id=acc["session_id"],
        session_file=str(latest_session_file) if latest_session_file else "",
        tool_call_counts=tool_call_counts,
        total_tool_calls=total_tool_calls,
        error_count=acc["error_count"],
        error_rate=error_rate,
        agent_spawns=acc["agent_spawns"],
        last_activity_at=acc["last_activity_at"],
        recent_timeline=timeline[-10:],
        input_tokens=acc["input_tokens"],
        output_tokens=acc["output_tokens"],
        cache_write_tokens=acc["cache_write_tokens"],
        cache_read_tokens=acc["cache_read_tokens"],
        total_tokens=total_tokens,
    )

File: skills/sortie/lib/parse-jsonl-metrics.mjs
/**
 * Parse Claude Code JSONL session logs to extract agent activity metrics.
 *
 * Session files live at:
 *   ~/.claude/projects/<encoded-path>/<session-uuid>.jsonl
 *   ~/.claude/projects/<encoded-path>/<session-uuid>/subagents/*.jsonl
 *
 * Path encoding: replace all '/' and '.' with '-'.
 * Each line is a JSON record. We care about:
 *   - type:"assistant" — contains tool_use blocks (tool calls) and usage data
 *   - type:"user"      — contains tool_result blocks (errors when is_error=true)
 */

import { createReadStream } from "node:fs";
import { readdir, stat } from "node:fs/promises";
import { join } from "node:path";
import { homedir } from "node:os";
import readline from "node:readline";

const CLAUDE_PROJECTS_DIR = join(homedir(), ".claude", "projects");

/** Safely coerce a potentially-malformed token count value to an integer. */
const si = (v) => (Number.isFinite(Number(v)) ? Math.trunc(Number(v)) : 0);

/**
 * Encode an absolute path the same way Claude Code does for its project dirs.
 * Both '/' and '.' are replaced with '-'.
 */
export function encodeProjectPath(absPath) {
  return absPath.replace(/[/.]/g, "-");
}

/**
 * Recursively find all .jsonl files under a directory (for subagent logs).
 */
async function findAllJsonlFiles(dir) {
  const files = [];
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      if (entry.isFile() && entry.name.endsWith(".jsonl")) {
        files.push(fullPath);
      } else if (entry.isDirectory()) {
        const nested = await findAllJsonlFiles(fullPath);
        files.push(...nested);
      }
    }
  } catch {}
  return files;
}

/**
 * Find the most recently modified top-level JSONL session file for a given worktree path.
 * Returns null if none found.
 */
export async function findLatestSessionFile(worktreePath) {
  const encoded = encodeProjectPath(worktreePath);
  const projectDir = join(CLAUDE_PROJECTS_DIR, encoded);

  let files;
  try {
    files = await readdir(projectDir);
  } catch {
    return null;
  }

  const jsonlFiles = files.filter((f) => f.endsWith(".jsonl"));
  if (jsonlFiles.length === 0) return null;

  const withStats = await Promise.all(
    jsonlFiles.map(async (f) => {
      const p = join(projectDir, f);
      try {
        const s = await stat(p);
        return { path: p, mtime: s.mtimeMs };
      } catch {
        return null;
      }
    })
  );

  const valid = withStats.filter(Boolean);
  if (valid.length === 0) return null;
  valid.sort((a, b) => b.mtime - a.mtime);
  return valid[0].path;
}

/**
 * Parse a single JSONL file and accumulate metrics into shared accumulators.
 *
 * @param {string} sessionFile
 * @param {Record<string, number>} toolCallCounts  mutated in place
 * @param {Array} timeline  mutated in place
 * @param {object} acc  mutated in place
 */
async function parseSingleFile(sessionFile, toolCallCounts, timeline, acc) {
  try {
    const rl = readline.createInterface({
      input: createReadStream(sessionFile, { encoding: "utf-8" }),
      crlfDelay: Infinity,
    });

    for await (const line of rl) {
      if (!line.trim()) continue;
      let obj;
      try {
        obj = JSON.parse(line);
      } catch {
        continue;
      }

      if (!acc.sessionId && obj.sessionId) {
        acc.sessionId = obj.sessionId;
      }

      if (obj.type === "assistant") {
        // Token usage — coerce to number to guard against malformed JSONL values
        const usage = obj.message?.usage ?? {};
        acc.inputTokens       += si(usage.input_tokens);
        acc.outputTokens      += si(usage.output_tokens);
        acc.cacheWriteTokens  += si(usage.cache_creation_input_tokens);
        acc.cacheReadTokens   += si(usage.cache_read_input_tokens);

        const content = obj.message?.content;
        if (Array.isArray(content)) {
          for (const block of content) {
            if (block?.type === "tool_use") {
              const name = block.name || "unknown";
              toolCallCounts[name] = (toolCallCounts[name] || 0) + 1;
              if (name === "Agent") acc.agentSpawns++;
              timeline.push({ tool: name, timestamp: obj.timestamp });
              acc.lastActivityAt = obj.timestamp;
            }
          }
        }
      }

      if (obj.type === "user") {
        const content = obj.message?.content;
        if (Array.isArray(content)) {
          for (const block of content) {
            if (block?.type === "tool_result" && block.is_error === true) {
              acc.errorCount++;
            }
          }
        }
      }
    }
  } catch {}
}

/**
 * Load JSONL metrics for a worktree path.
 * Scans ALL session files (including subagent logs) for token usage.
 * Returns null if no session files exist.
 *
 * @param {string} worktreePath
 * @returns {Promise<JsonlMetrics|null>}
 */
export async function parseJsonlMetrics(worktreePath) {
  const encoded = encodeProjectPath(worktreePath);
  const projectDir = join(CLAUDE_PROJECTS_DIR, encoded);

  const allFiles = await findAllJsonlFiles(projectDir);
  if (allFiles.length === 0) return null;

  // Latest top-level session for metadata
  const topLevelFiles = allFiles.filter(
    (f) => f.startsWith(projectDir) && !f.slice(projectDir.length + 1).includes("/")
  );
  let latestSessionFile = null;
  if (topLevelFiles.length > 0) {
    const withStats = await Promise.all(
      topLevelFiles.map(async (f) => {
        try { return { path: f, mtime: (await stat(f)).mtimeMs }; }
        catch { return null; }
      })
    );
    const valid = withStats.filter(Boolean);
    if (valid.length > 0) {
      valid.sort((a, b) => b.mtime - a.mtime);
      latestSessionFile = valid[0].path;
    }
  }

  /** @type {Record<string, number>} */
  const toolCallCounts = {};
  /** @type {Array<{tool: string, timestamp: string}>} */
  const timeline = [];
  const acc = {
    sessionId: null,
    inputTokens: 0,
    outputTokens: 0,
    cacheWriteTokens: 0,
    cacheReadTokens: 0,
    agentSpawns: 0,
    errorCount: 0,
    lastActivityAt: null,
  };

  // Parse latest session file first so sessionId comes from the correct session
  if (latestSessionFile) {
    await parseSingleFile(latestSessionFile, toolCallCounts, timeline, acc);
  }
  for (const file of allFiles) {
    if (file === latestSessionFile) continue;
    await parseSingleFile(file, toolCallCounts, timeline, acc);
  }

  // Sort timeline by timestamp so cross-file ordering is correct, then derive
  // lastActivityAt from the true latest event rather than traversal order.
  // Parse to epoch so ISO 8601 timestamps with offsets sort correctly.
  timeline.sort((a, b) => {
    const ta = a.timestamp ? (Date.parse(a.timestamp) || 0) : 0;
    const tb = b.timestamp ? (Date.parse(b.timestamp) || 0) : 0;
    return ta - tb;
  });
  if (timeline.length > 0) {
    acc.lastActivityAt = timeline[timeline.length - 1].timestamp;
  }

  const totalToolCalls = Object.values(toolCallCounts).reduce((a, b) => a + b, 0);
  const errorRate = totalToolCalls > 0
    ? Math.round((acc.errorCount / totalToolCalls) * 100) / 100
    : 0;

  const totalTokens = acc.inputTokens + acc.outputTokens + acc.cacheWriteTokens + acc.cacheReadTokens;

  return {
    sessionId: acc.sessionId,
    sessionFile: latestSessionFile,
    toolCallCounts,
    totalToolCalls,
    errorCount: acc.errorCount,
    errorRate,
    agentSpawns: acc.agentSpawns,
    lastActivityAt: acc.lastActivityAt,
    recentTimeline: timeline.slice(-10),
    inputTokens: acc.inputTokens,
    outputTokens: acc.outputTokens,
    cacheWriteTokens: acc.cacheWriteTokens,
    cacheReadTokens: acc.cacheReadTokens,
    totalTokens,
  };
}
